django/db/models/fields/related.py

```diff
@@ -1030,8 +1036,59 @@ def _check_on_delete(self):
                     id="fields.E321",
                 )
             ]
-        else:
-            return []
+        elif (
+            on_delete in [DB_CASCADE, DB_SET_NULL]
+            and hasattr(self.model, "_meta")
+            and any(
+                not parent._meta.abstract
+                for parent in self.model._meta.get_parent_list()
+            )
+        ):
+            return [
+                checks.Error(
+                    "Field specifies unsupported on_delete=DB_CASCADE, on "
+                    "inherited model as it already contains a parent_ptr "
+                    "with in-python cascading options, both cannot be used "
+                    "together",
+                    hint="Change the on_delete rule to other options",
+                    obj=self,
+                    id="fields.E325",
+                )
+            ]
+        elif (
+            on_delete in [DB_CASCADE, DB_SET_NULL]
+            and hasattr(self.model, "_meta")
+            and (
+                any(  # generic relation
+                    hasattr(field, "bulk_related_objects")
+                    for field in self.model._meta.private_fields
+                )
+                or any(  # generic foreign key
+                    hasattr(field, "get_content_type")
+                    for field in self.model._meta.private_fields
+                )
+            )
+        ):
+            return [
+                checks.Error(
+                    "Field specifies unsupported on_delete=DB_CASCADE on model "
+                    "declaring a GenericForeignKey.",
```

Comment:
It might be useful to specify which `field` is the incompatible generic relation or generic foreign key.

Reply:
Sounds good, will update the patch accordingly.

Reply:
The field name is being passed, so I think this will not be necessary
```
db_level_on_delete_checks.SomeAnotherModel.field_name: (fields.E345) Field specifies unsupported on_delete=DB_CASCADE on model declaring a GenericForeignKey.
        HINT: Change the on_delete rule.
```